plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java

```diff
@@ -303,13 +313,62 @@ private JdbcTableHandle flushAttributesAsQuery(ConnectorSession session, JdbcTab
         handle = new JdbcTableHandle(
                 handle.getRelationHandle(),
                 handle.getConstraint(),
+                handle.getSortOrder(),
                 OptionalLong.of(limit),
                 handle.getColumns(),
                 handle.getNextSyntheticColumnId());
 
         return Optional.of(new LimitApplicationResult<>(handle, jdbcClient.isLimitGuaranteed(session)));
     }
 
+    @Override
+    public Optional<TopNApplicationResult<ConnectorTableHandle>> applyTopN(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            long topNCount,
+            List<SortItem> sortItems,
+            Map<String, ColumnHandle> assignments)
+    {
+        if (!isTopNPushdownEnabled(session)) {
+            return Optional.empty();
+        }
+
+        verify(!sortItems.isEmpty(), "sortItems are empty");
+        JdbcTableHandle handle = (JdbcTableHandle) table;
+
+        List<SortItem> resultSortOrder = sortItems.stream()
+                .map(sortItem -> {
+                    verify(assignments.containsKey(sortItem.getName()), "assignments does not contain sortItem: %s", sortItem.getName());
+                    return new SortItem(((JdbcColumnHandle) assignments.get(sortItem.getName())).getColumnName(), sortItem.getSortOrder());
+                })
+                .collect(toImmutableList());
+
+        if (!jdbcClient.supportsTopN(session, handle, assignments, resultSortOrder)) {
+            // JDBC client implementation prevents TopN pushdown for the given table
+            // i.e. when order by on a given type does not match Trino semantic
+            return Optional.empty();
+        }
+
+        if (handle.getSortOrder().isPresent() || handle.getLimit().isPresent()) {
+            if (handle.getLimit().equals(OptionalLong.of(topNCount)) && handle.getSortOrder().equals(Optional.of(resultSortOrder))) {
+                return Optional.of(new TopNApplicationResult<>(handle, jdbcClient.isTopNLimitGuaranteed(session)));
```

Comment:
I think we should return `Optional.empty` here, per
https://github.com/trinodb/trino/blob/9957543e57aa10b094ac7a7a55b0df7779d61125/core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java#L1067-L1069

returning a `TopNApplicationResult` will cause the optimize to loop when `isTopNLimitGuaranteed = false`, but IMO we should return `empty` regardless.

Reply:
That breaks the code so it doesn't work anymore

Reply:
What breaks? i didn't get this one

Reply:
If I return `Optional.empty()` there, assertions `isFullyPushedDown` will fail as top-level `TopN` and `Exchange` nodes will remain in a plan.

Reply:
We need to see whether we can defer creation of partial TopN, so that we don't need to deal with partial/final TopN in the TopN pushdown.
Then the engine rule would be narrowed down to SINGLE TopN s only. 
Do you feel like you could investigate this?

Reply:
I've applied changes to the optimizers and PushTopNIntoTableScan rule. PTAL now.